### PR TITLE
fix(plugin-pi): accept underscore-form Remnic MCP tool names

### LINKED ISSUE #2788: plugin-pi: MCP tool-name mismatch — server returns remnic_* (underscore) names, plugin filters on remnic. (dotted), so no tools register
## Summary

The deployed Remnic server and official `@remnic/plugin-pi@9.69.4` disagree about canonical MCP tool names. The server's authenticated `tools/list` returns **underscore-form** names (`remnic_recall`, `remnic_recall_explain`, …), but the plugin filters discovered tools using:

```ts
if (!tool.name.startsWith("remnic.")) continue;
```

It expects **dotted** server names (`remnic.recall`) and converts them into Pi-safe registered names (`remnic.recall` → `remnic_recall`). Because the server already returns underscore names, every Remnic tool is rejected before `pi.registerTool()` is called, so no top-level Remnic tools are ever registered.

## Symptom: confusing partial success

- Automatic recall/observe hooks still work through the direct `/engram/v1/*` endpoints.
- Namespace preflight can still succeed.
- **No top-level `remnic_*` tools are registered.**
- The agent therefore reports that `remnic_recall` and related tools are unavailable, even though Remnic itself appears healthy.

This silent empty-catalog failure gives no actionable diagnostic.

## Evidence

An authenticated JSON-RPC `tools/list` call against the hosted Remnic server returns 101 underscore-form tools:

```text
remnic_recall
remnic_recall_explain
remnic_set_coding_context
… (98 more, all in remnic_* form)
```

The plugin's filter (`startsWith("remnic.")`) rejects all 101.

## Confirmed workaround / end-to-end reproduction

Reproduced end-to-end in OMP (which consumes the official Pi extension API):

1. Native extension loaded; hook behavior remained active.
2. Authenticated direct `tools/list` returned 101 underscore-form Remnic tools.
3. A real OMP session could not call `remnic_recall`.
4. A temporary compatibility adapter translated:
   - inbound catalog names: `remnic_recall` → `remnic.recall`
   - outbound tool calls: `remnic.recall` → `remnic_recall`
5. After translation, a real, normally-started OMP session successfully registered and called the top-level `remnic_recall` tool.

This proves the name mismatch is the sole cause.

## Version observed

`@remnic/plugin-pi@9.69.4`. Current upstream `main` should be re-checked before fixing — the bug may have changed since that published version.

## Expected behavior

- `@remnic/plugin-pi` must support the canonical underscore names currently returned by the Remnic server.
- Preserve compatibility with older dotted names (`remnic.recall`) if they remain supported by any released server.
- Do not blindly replace every underscore or period in arbitrary MCP tool names; normalize only the recognized Remnic namespace/prefix.
- The name sent in `tools/call` must be accepted by the server that produced the catalog (i.e., echo back the catalog name, not a re-derived one).
- No toolbox/OMP-specific fetch interception; the fix belongs in shared plugin behavior.

## Diagnostics

If `tools/list` succeeds but yields zero acceptable Remnic tools, the plugin must surface actionable information instead of silently presenting an empty catalog.

## Acceptance criteria

1. An underscore-form catalog (`remnic_recall`, `remnic_recall_explain`, `remnic_set_coding_context`) registers the expected Pi-safe tools.
2. A legacy dotted catalog (`remnic.recall`, `remnic.recall_explain`, `remnic.set_coding_context`) continues working if dotted names remain part of the compatibility contract.
3. Calling a registered tool sends the correct server-side tool name (tested for both catalog shapes if both are supported).
4. Non-Remnic tools remain excluded from the registration loop.
5. Malformed names or unrelated names that merely contain "remnic" are not normalized accidentally.
6. Empty-catalog-after-successful-discovery produces a visible diagnostic.
7. Existing automatic recall, observe, sessionKey, namespace, and cwd injection tests continue passing.

### PR DESCRIPTION
**Pull request by ramarivera**
Fixes #2788

## Summary

`@remnic/plugin-pi` filters discovered MCP tools with `tool.name.startsWith("remnic.")`, but current Remnic servers return **underscore-form** canonical names (`remnic_recall`, `remnic_recall_explain`, …) from authenticated `tools/list`. Every Remnic tool was therefore rejected before `pi.registerTool()`, while hooks kept working through the direct `/engram/v1/*` endpoints — a confusing partial-success state where the agent reports `remnic_recall` as unavailable despite a healthy Remnic connection. Full evidence and reproduction in #2788.

## Changes (focused, no unrelated refactors)

- Add exported `matchRemnicMcpToolName()` helper that normalizes **only** the recognized Remnic namespace prefixes:
  - legacy dotted: `remnic.recall` → Pi-safe `remnic_recall`
  - current underscore: `remnic_recall` → `remnic_recall`
  - arbitrary names merely containing "remnic" (`myremnic_recall`, `prefix_remnic_recall`), bare prefixes (`remnic.`, `remnic_`), and non-string values are rejected
- Registered tools echo the **catalog name verbatim** in `tools/call`, so requests are always accepted by whichever server shape produced the catalog.
- Recall-timeout-breaker detection (`remnic.recall` / `remnic.recall_xray`) now works for both catalog shapes.
- Mixed-shape catalogs dedupe onto one Pi-safe registration per tool (first occurrence wins).
- Diagnostics: if `tools/list` succeeds but zero tools match, the plugin logs a `console.warn` and appends a structured `remnic_state` entry (`code: "MCP_TOOLS_FILTERED"`) listing observed names — no more silent empty catalogs.
- No toolbox/OMP-specific interception; the fix lives entirely in shared plugin behavior.

## Tests

Focused regression tests in `packages/plugin-pi/src/index.test.ts`:

1. Underscore catalog (`remnic_recall`, `remnic_recall_explain`, `remnic_set_coding_context`) registers expected Pi-safe tools.
2. Legacy dotted catalog (`remnic.recall`, …) continues to register and call the dotted server-side name.
3. Registered tools send the correct server-side name in `tools/call` (both catalog shapes).
4. Non-Remnic tools remain excluded from the registration loop.
5. Malformed/unrelated names merely containing "remnic" are not normalized; empty-catalog-after-successful-discovery emits both console and structured diagnostics.
6. Mixed-shape catalog dedupe.
7. Unit coverage for `matchRemnicMcpToolName` edge cases.

Validation:

- `pnpm install --filter @remnic/plugin-pi...`
- `npm run check-types` in `packages/plugin-pi` → clean (`tsc --noEmit`)
- `npm test` in `packages/plugin-pi` → **176 pass / 0 fail**, including all pre-existing automatic recall, observe, sessionKey, namespace, and cwd injection tests

## Release notes

Added changeset `.changeset/2788-pi-omp-tools-filtering.md` (`@remnic/plugin-pi`: patch).

Observed on `@remnic/plugin-pi@9.69.4`; verified still present on current upstream `main` before fixing.

## Summary by CodeRabbit

- **New Features**
  - Added support for Remnic MCP tools using dotted or underscore-based names.
  - Preserved original tool names when executing MCP calls.
  - Deduplicated tools with equivalent naming formats.
  - Added diagnostics and warnings when no matching tools are discovered.

- **Bug Fixes**
  - Improved recall tool compatibility across supported naming formats.
  - Excluded unrelated, malformed, or punctuation-containing tool names from registration.

### CODE REVIEW

File: packages/plugin-pi/src/index.test.ts [hunk 1]
```diff
@@ -1567,6 +1568,168 @@ test("remnic-why remains available after the recall timeout breaker trips", asyn
...
+    },
+  });
+  await extension({
+    on: () => undefined,
+    registerCommand: () => undefined,
+    registerTool: (tool: Record<string, unknown>) => registeredTools.push(tool as RegisteredTool),
+    appendEntry: (customType: string, data?: unknown) => appendedEntries.push({ customType, data }),
+  });
+
+  return { registeredTools, calledToolNames, appendedEntries };
```
[the remark this answers is not in the document]
Comment (ramarivera) [reply packages/plugin-pi/src/index.test.ts]: Fixed in 2fa4c86. `registerToolsFromCatalog` now keeps all mock-dependent setup and tool execution inside `try/finally`, restoring `globalThis.fetch` before it returns. Focused plugin Pi tests pass (63/63).

File: packages/plugin-pi/src/index.test.ts [hunk 2]
```diff
@@ -1567,6 +1568,168 @@ test("remnic-why remains available after the recall timeout breaker trips", asyn
...
+  try {
+    const { registeredTools, appendedEntries } = await registerToolsFromCatalog([
+      "other_server_tool",
+      "myremnic_recall",
+      "prefix_remnic_recall",
+      "remnic",
+      "remnic.",
+      "remnic_",
+    ]);
+    assert.deepEqual(registeredTools, []);
```
[the remark this answers is not in the document]
Comment (ramarivera) [reply packages/plugin-pi/src/index.test.ts]: Fixed in 2fa4c86. The malformed-catalog test now asserts `registeredTools.length === 0` rather than comparing directly with an empty array. Focused plugin Pi tests pass (63/63).

File: packages/plugin-pi/src/index.ts [hunk 3]
```diff
@@ -50,6 +50,41 @@ const RECALL_PRODUCING_TOOL_NAMES: Record<string, true> = {
...
+  if (name.startsWith("remnic.")) {
+    rest = name.slice("remnic.".length);
+  } else if (name.startsWith("remnic_")) {
+    rest = name.slice("remnic_".length);
+  }
+  if (rest === null || rest.length === 0) return null;
+  return {
+    serverToolName: name,
+    piToolName: `remnic_${rest}`.replace(/[^a-zA-Z0-9_]/g, "_"),
+  };
```
[the remark this answers is not in the document]
Comment (ramarivera) [reply packages/plugin-pi/src/index.ts]: Fixed in 2fa4c86. The matcher now rejects nonempty suffixes containing characters outside `[A-Za-z0-9_]` before deriving the Pi name; regression coverage includes `remnic_!` and `remnic.recall!`. Focused plugin Pi tests pass (63/63).

File: packages/plugin-pi/src/index.test.ts [hunk 4]
```diff
@@ -1567,6 +1568,168 @@ test("remnic-why remains available after the recall timeout breaker trips", asyn
...
+  });
+  await extension({
+    on: () => undefined,
+    registerCommand: () => undefined,
+    registerTool: (tool: Record<string, unknown>) => registeredTools.push(tool as RegisteredTool),
+    appendEntry: (customType: string, data?: unknown) => appendedEntries.push({ customType, data }),
+  });
+
+  return { registeredTools, calledToolNames, appendedEntries };
+}
```
[the remark this answers is not in the document]
Comment (ramarivera) [reply packages/plugin-pi/src/index.test.ts]: Fixed in 2fa4c86. This is the same fetch-restoration issue also reported by CodeRabbit; the helper now restores `globalThis.fetch` in `finally` after its mock-dependent execution. Focused plugin Pi tests pass (63/63).
